File: solutions/python/triangle/1/triangle.py
def is_triangle(sides):
    a,b,c = sides
    return (a+b+c > 0) and (a + b > c) and (a + c > b) and (b + c > a)

def equilateral(sides):
    a,b,c = sides
    return a==b and b==c and is_triangle(sides)


def isosceles(sides):
    a,b,c = sides
    return (a==b or b==c or a==c) and is_triangle(sides)


def scalene(sides):
    return not (isosceles(sides) or equilateral(sides)) and is_triangle(sides)

File: solutions/javascript/triangle/1/triangle.js
export class Triangle {
  constructor(...sides) {
    if (sides.length != 3)
    {
      throw new Error('Not a triangle');
    }
    // check sides are strict positive numbers
    // and that the triangle is possible
    // (ie: for instance there is no triangle with sides 3, 5, 10)

    this.sides = sides.sort();
  }

  get isEquilateral() {
    // this is merely to make the test pass
    // it would be better to detect all non triangles
    // at constructor raising some exception
    // at it's not the API picked by tests
    // I won't care writing the sanity check code everywhere
    let [a, b, c] = this.sides;
    if (a + b + c == 0){
      return false;
    }
    return a == b && b == c;
  }

  get isIsosceles() {
    // if the triangle is equilateral it is also isocele
    let [a, b, c] = this.sides;
    if (a + b < c) {
      return false;
    }
    return a == b || b == c || a == c ;
  }

  get isScalene() {
    let [a, b, c] = this.sides;
    if (a + b < c) {
      return false;
    }
    return !this.isIsosceles;
  }
}
